Author: Guy, Michael, Joshua
Summary: This is a comprenhensive simulation engine for ambulance simulation. It uses the solution from two-stage stochastic program and robust programs to simulate how ambulances would perform. The simulation mostly runs using a priority-queue of events with respect to time. Events include things like ambulance arrives to the scene of call, ambulance arrives to hospital, ect.

In [32]:
import Pkg #this line doesn't seem to run either way for me
Pkg.add("Pkg")
Pkg.add("Dates")

  Resolving package versions...
No Changes to `C:\Users\Owner\Documents\Austin\AmbulanceDeployment\AmbulanceDeployment.jl-legacy\Project.toml`
No Changes to `C:\Users\Owner\Documents\Austin\AmbulanceDeployment\AmbulanceDeployment.jl-legacy\Manifest.toml`
  Resolving package versions...
No Changes to `C:\Users\Owner\Documents\Austin\AmbulanceDeployment\AmbulanceDeployment.jl-legacy\Project.toml`
No Changes to `C:\Users\Owner\Documents\Austin\AmbulanceDeployment\AmbulanceDeployment.jl-legacy\Manifest.toml`


In [34]:
"""This activates packages, namely ambulance deployment"""

guypath = "/Users/guyfarmer/ambulance_connect/AmbulanceDeployement/AmbulanceDeployment.jl-legacy/"
michaelpath= "/home/mhilborn/Desktop/UT-Spring2021/SeniorLab/AmbulanceDeployement/AmbulanceDeployment.jl-legacy"
joshpath = "C:/Users/Owner/Documents/Austin/AmbulanceDeployment/AmbulanceDeployment.jl-legacy/"

currentpath = joshpath

cd(currentpath)
Pkg.activate(".")

using DataFrames, JLD, Distributions, CSV, Random, Plots,JSON, DelimitedFiles
import DataStructures: PriorityQueue, enqueue!, dequeue!
include(".//src//problem.jl")
include(".//src//model.jl")

using AmbulanceDeployment, Dates

 Activating environment at `C:\Users\Owner\Documents\Austin\AmbulanceDeployment\AmbulanceDeployment.jl-legacy\Project.toml`


In [36]:
"""Helper functions for the simuation engine?"""
#=
Author : Ng Yeesian
Modified : Guy Farmer / Zander Tedjo / Will Worthington / Michael Hilborn
generates the closest dispatch model
Summary :
Input :
Output :
=#
mutable struct ClosestDispatch <: DispatchModel
    drivetime::DataFrame
    candidates::Vector{Vector{Int}}
    available::Vector{Int}
    assignment::Vector{Int} # which location the ambulance is assigned to
    ambulances::Vector{Vector{Int}} # list of ambulances assigned to each location
    status::Vector{Symbol} # the current status of the ambulance
        # possible statuses: :available, :responding, :atscene, :conveying, :returning
    fromtime::Vector{Int} # the time it started the new status
    hospital::Vector{Int}
end

#=
Summary :
Input :
Output : ClosestDispatch struct
=#
function ClosestDispatch(p::DeploymentProblem, drivetime::DataFrame,distribution::Vector{Int})
    candidates = Any[]
    I = 1:p.nlocations
    for region in 1:p.nregions
        push!(candidates, I[vec(p.coverage[region,:])])
    end
    #println("creating the dispatch problem :$distribution")

    assignment = zeros(Int, p.nambulances)
    ambulances = [Int[] for i in 1:p.nlocations]
    k = 1
    for i in eachindex(distribution), j in 1:distribution[i]
        assignment[k] = i
        push!(ambulances[i], k)
        k += 1
    end
    status = fill(:available, p.nambulances)
    fromtime = zeros(Int, p.nambulances)
    hospital = zeros(Int, p.nambulances)
    
    ClosestDispatch(drivetime, candidates, distribution, assignment, ambulances, status, fromtime, hospital)

end

function ClosestDispatch(p::DeploymentProblem, problem::DispatchProblem)
    stn_names = [Symbol("stn$(i)_min") for i in 1:size(p.coverage,2)]
    ClosestDispatch(p, problem.emergency_calls[:, stn_names], problem.available)
end

function update_ambulances!(model::ClosestDispatch, i::Int, delta::Int)
    model.available[i] += delta
    @assert model.available[i] >= 0
end

function closest_available(dispatch::ClosestDispatch, id::Int, problem::DispatchProblem)
    location = 0
    min_time = Inf
    for i in dispatch.candidates[problem.emergency_calls[id, :neighborhood]]
        if problem.available[i] > 0 && dispatch.drivetime[id, i] < min_time
            #println("apparently this station is available : $i and here is the amount of ambulances avalable: $(problem.available[i])")
            location = i
            min_time = dispatch.drivetime[id, i]
        end
    end
    location
end

function respond_to!(dispatch::ClosestDispatch, i::Int, t::Int)
    @assert length(dispatch.ambulances[i]) > 0 "$i: $(dispatch.ambulances[i])"
    amb = popfirst!(dispatch.ambulances[i])
    # @assert redeploy.hospital[amb] == 0
    @assert amb != 0
    @assert dispatch.status[amb] == :available "$amb: $(dispatch.status[amb])"
    dispatch.status[amb] = :responding
    #dispatch.fromtime[amb] = t
    amb
end

function arriveatscene!(dispatch::ClosestDispatch, amb::Int, t::Int)
    @assert dispatch.status[amb] == :responding "$amb: $(dispatch.status[amb])"
    @assert dispatch.hospital[amb] == 0
    dispatch.status[amb] = :atscene
    dispatch.fromtime[amb] = t
end

function arriveathospital!(dispatch::DispatchModel, amb::Int, hosp::Int, t::Int)
    @assert dispatch.status[amb] == :atscene "$amb: $(dispatch.status[amb])"
    @assert dispatch.hospital[amb] != 0
    dispatch.status[amb] = :atHospital
    dispatch.fromtime[amb] = t
    dispatch.hospital[amb] = hosp
end

function returning_to!(dispatch::DispatchModel, amb::Int, t::Int)
    @assert dispatch.status[amb] == :atHospital "$amb: $(dispatch.status[amb])"
    @assert dispatch.hospital[amb] != 0
    dispatch.status[amb] = :returning
    dispatch.fromtime[amb] = t
    dispatch.assignment[amb]
end

returning_to! (generic function with 1 method)

In [49]:
""" simulation parameters """
turnaround = Distributions.LogNormal(3.65, 0.3)#Distributions.LogNormal(3.089, 0.3) #Distributions.LogNormal(3.65, 0.3)
lambda = 0
local_path = ""
model_name = "Stochastic"
namb = 40
ncalls = 1000
#using AmbulanceDeployment

pwd()

cd(currentpath * "/test/")
pwd()

#using AmbulanceDeployment
using Dates

#include("..//src//dispatch/closestdispatch.jl")
include("..//src//problem.jl")
#include("..//src//evaluate.jl") #if you include, you need to include JuMP
#include("..//src//simulate.jl")

hospitals = CSV.File(string(local_path,"../test/austin-data/hospitals.csv")) |> DataFrame
stations = CSV.File(string(local_path,"../test/austin-data/stations.csv")) |> DataFrame
hourly_calls = CSV.File("../test/austin-data/Full_WeekdayCalls.csv") |> DataFrame
adjacent_nbhd = CSV.File("../test/austin-data/adjacent_nbhd.csv") |> DataFrame
coverage = CSV.read("../test/austin-data/coverage_real.csv", DataFrame, header=false)
coverage = convert(Array{Bool, 2}, coverage[:, :])
incidents = CSV.File("../test/austin-data/austin_incidents.csv") |> DataFrame
solverstats = JLD.load("../src/outputs/austin_team_stats_1_03.jld")
amb_deployment = solverstats["amb_deployment"]
model_dict = Dict{String, Symbol}("Stochastic"=>:Stochastic, "Robust01"=>:Robust01, "Robust005"=>:Robust005, "Robust001"=>:Robust001, "Robust0001"=>:Robust0001,
          "Robust00001"=>:Robust00001, "MEXCLP"=>:MEXCLP, "MALP"=>:MALP)


p = DeploymentProblem(
          hourly_calls,
          adjacent_nbhd,
          coverage,
          namb = namb,
          train_filter = (hourly_calls[!,:year] .== 2019) .* (hourly_calls[!,:month] .<= 3)
          )
test_calls = CSV.File("../test/austin-data/austin_test_calls_v1.csv")|> DataFrame 
#update 1: remove negative indices
test_calls = test_calls[1:ncalls,:]
println("running $namb ambuances & $ncalls calls")

# distribution of ambulances across all of the stations
x = amb_deployment[model_dict[model_name]][namb]

#establishes the dispatch problem
problem = DispatchProblem(test_calls, hospitals, stations, p.coverage, x, turnaround=turnaround)

#establishes the dispatch model
dispatch = ClosestDispatch(p, problem)

Random.seed!(1234)



running 40 ambuances & 1000 calls


MersenneTwister(UInt32[0x000004d2], Random.DSFMT.DSFMT_state(Int32[-1393240018, 1073611148, 45497681, 1072875908, 436273599, 1073674613, -2043716458, 1073445557, -254908435, 1072827086  …  -599655111, 1073144102, 367655457, 1072985259, -1278750689, 1018350124, -597141475, 249849711, 382, 0]), [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0  …  0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0], UInt128[0x00000000000000000000000000000000, 0x00000000000000000000000000000000, 0x00000000000000000000000000000000, 0x00000000000000000000000000000000, 0x00000000000000000000000000000000, 0x00000000000000000000000000000000, 0x00000000000000000000000000000000, 0x00000000000000000000000000000000, 0x00000000000000000000000000000000, 0x00000000000000000000000000000000  …  0x00000000000000000000000000000000, 0x00000000000000000000000000000000, 0x00000000000000000000000000000000, 0x00000000000000000000000000000000, 0x00000000000000000000000000000000, 0x00000000000000000000000000000000, 0x000000000000

In [51]:
test_calls

1000×60 DataFrame. Omitted printing of 55 columns
│ Row  │ interarrival_seconds │ neighborhood │ dow    │ stn1_min │ stn2_min │
│      │ Int64                │ Int64        │ String │ Float64  │ Float64  │
├──────┼──────────────────────┼──────────────┼────────┼──────────┼──────────┤
│ 1    │ 38                   │ 83           │ Wed    │ 5.0      │ 26.9957  │
│ 2    │ 537                  │ 38           │ Wed    │ 16.873   │ 26.3413  │
│ 3    │ 98                   │ 129          │ Wed    │ 21.5312  │ 38.0153  │
│ 4    │ 18                   │ 107          │ Wed    │ 44.2525  │ 37.6413  │
│ 5    │ 475                  │ 113          │ Wed    │ 14.653   │ 29.5155  │
│ 6    │ 21                   │ 51           │ Wed    │ 21.06    │ 13.8817  │
│ 7    │ 263                  │ 76           │ Wed    │ 43.5517  │ 34.0775  │
│ 8    │ 128                  │ 67           │ Wed    │ 14.8368  │ 17.8243  │
│ 9    │ 62                   │ 52           │ Wed    │ 17.9115  │ 22.8192  │
│ 10   │ 198                  │ 67           │ Wed    │ 14.8368  │ 17.8243  │
⋮
│ 990  │ 799                  │ 85           │ Tue    │ 19.207   │ 34.1975  │
│ 991  │ 92                   │ 83           │ Tue    │ 5.0      │ 26.9957  │
│ 992  │ 520                  │ 128          │ Tue    │ 20.2293  │ 30.526   │
│ 993  │ 148                  │ 128          │ Tue    │ 20.2293  │ 30.526   │
│ 994  │ 742                  │ 113          │ Tue    │ 14.653   │ 29.5155  │
│ 995  │ 62                   │ 97           │ Tue    │ 14.0803  │ 19.063   │
│ 996  │ 1291                 │ 113          │ Tue    │ 14.653   │ 29.5155  │
│ 997  │ 428                  │ 99           │ Tue    │ 12.6037  │ 35.743   │
│ 998  │ 648                  │ 116          │ Tue    │ 20.3053  │ 38.0978  │
│ 999  │ 2905                 │ 97           │ Tue    │ 14.0803  │ 19.063   │
│ 1000 │ 913                  │ 82           │ Tue    │ 10.3225  │ 19.6     │

In [53]:
problem.emergency_calls[:, Symbol("stn$(1)_min")]

1000-element Array{Float64,1}:
  5.0
 16.873
 21.531166666666664
 44.252500000000005
 14.652999999999999
 21.06
 43.55166666666666
 14.836833333333335
 17.9115
 14.836833333333335
 14.836833333333335
 10.3225
 10.132
  ⋮
 20.229333333333333
 19.207
  5.0
 20.229333333333333
 20.229333333333333
 14.652999999999999
 14.080333333333334
 14.652999999999999
 12.603666666666667
 20.305333333333333
 14.080333333333334
 10.3225

In [55]:
function gridTime2Bin(gridTime)
    weights = [201.70483660130716, 10.166325878594247, -208.83953125000002, -460.87083333333334,-775.0755555555556,-1046.168888888889,-1527.7117647058826]
    bins = [ 250,  550,  850, 1150, 1450, 1750]
    for i in 1:length(bins)
        if(gridTime < bins[i])
            return (i)
        end
    return length(bins) + 1
    end
end

gridTime2Bin (generic function with 1 method)

In [62]:
"""Simulation Engine"""

Pkg.add("Query")
using Query

struct EMSEngine{T}
    eventlog::DataFrame
    eventqueue::PriorityQueue{T,Int,Base.Order.ForwardOrdering}
    shortqueue::PriorityQueue{T,Int,Base.Order.ForwardOrdering}
    guiArray::Array{Any, 1}
end

struct gui_event
    #call made / call responded / ambulance arrived / shortfall
    event_type::String
    #neighborhood where call is made from
    neighborhood_id::Int
    # station where call is made from
    deployment_id::Int
    # id of the event generated
    event_id::Int
    ambulance_id::Int
    arrival_time::Int
    remaining_amb::Int
    # current timestamp of event
    timestamp::DateTime
end

## generates the priority queue and instantiates the EMSEngine struct
function generateCalls(problem::DispatchProblem)
    ncalls = nrow(problem.emergency_calls)
    eventlog = DataFrame(
        id = 1:ncalls,
        dispatch_from = zeros(Int, ncalls),
        waittime = fill(0.0, ncalls),
        responsetime = fill(Inf, ncalls),
        scenetime = fill(Inf, ncalls),
        conveytime = fill(Inf, ncalls),
        returntime = fill(Inf, ncalls),
        return_to = zeros(Int, ncalls),
        return_type = fill(:station, ncalls),
        hospital = zeros(Int, ncalls),
        ambulance = zeros(Int, ncalls)
    )
    eventqueue = PriorityQueue{Tuple{Symbol,Int,Int,Int},Int}()
    shortqueue = PriorityQueue{Tuple{Symbol,Int,Int,Int},Int}()
    guiArray = Any[]
    for i in 1:nrow(problem.emergency_calls)
        t = problem.emergency_calls[i, :arrival_seconds]
        enqueue!(eventqueue, (:call, i, t, problem.emergency_calls[i, :neighborhood]), t)
    end
    EMSEngine{Tuple{Symbol,Int,Int,Int}}(eventlog, eventqueue, shortqueue, guiArray) ## returns the ems engine struct
end

function simulate_events!(problem::DispatchProblem, dispatch::ClosestDispatch)
    engine = generateCalls(problem)
    while(~isempty(engine.eventqueue))
        (event, id, t, value) = dequeue!(engine.eventqueue)

        if event == :call
            # if the event is reachable then change event to station2call
            # if not it is a shortfall
            println("calling event id: $id time: $t value: $value")
            call_event!(engine, problem, dispatch, id, t, value)
           
        elseif event == :station2call
            #once the ambulance arrives at the station change event to call2hospital
            println("Ambulance has arrived on the scene for event $id at time $t")
            station2call_event!(engine, problem, dispatch, id, t, value)
        elseif event == :call2hospital
            #once the ambulance arrives at the hospital it will hand over the patient and then 
            #change the deploy back to its station
            println("Ambulance for event $id has arrived at the hospital at time $t")
            call2hosp_event!(engine, problem, dispatch, id, t, value)

        elseif event == :hospital2station
            #once the ambulance drops off the patient it will reroute to the station
            println("Ambulance for event $id is returning to the station at $t")
            return_event!(engine, problem, dispatch, id, t, value)
            
        elseif event == :done
            #increase the number of ambulances at that station now that it is done
            done_event!(engine, problem, dispatch, id, t, value)
            
        end
    end
    println("This is the Short Queue")
    while(~isempty(engine.shortqueue))
       (event, id, t, value) = dequeue!(engine.shortqueue)
        print("$event, $id, $t, $value")
        print("This is not empty")
    end
    engine.eventlog , engine.guiArray
end


function call_event!(
        ems::EMSEngine,
        problem::DispatchProblem,
        dispatch::DispatchModel,
        id::Int, # the id of the emergency call
        t::Int, # the time of the emergency call
        nbhd::Int; # the neighborhood the call is from
    )

    #check if there is an ambulance within the coverage matrix
    if sum(problem.deployment[problem.coverage[nbhd,:]]) == 0
        @assert false "$id: no ambulance reachable for call at $nbhd"
    #check if one of the ambulances within the coverage is available
    elseif sum(problem.available[problem.coverage[nbhd,:]]) > 0
        i = closest_available(dispatch, id, problem)
        #println("closest available station : $i")
        @assert i > 0 "$id: dispatch from $i to nbhd $nbhd" # assume valid i (enforced by <if> condition)
        update_ambulances!(dispatch, i, -1)
        ems.eventlog[id, :dispatch_from] = i
        @assert problem.available[i] >= 0
        

        travel_time = ceil(Int, 60*problem.emergency_calls[id, Symbol("stn$(i)_min")])
        travel_time = travel_time -146 #+ gridTime2Bin(travel_time) #adjust time
        #travel_time = ceil(travel_time * .9) #time mod
        @assert travel_time >= 0
        if(t != problem.emergency_calls[id, :arrival_seconds])
            print("This is a shortfall")
        end
        ems.eventlog[id, :responsetime] = (travel_time + t - problem.emergency_calls[id,:arrival_seconds]) / 60 # minutes

        amb = respond_to!(dispatch, i, t)
        ems.eventlog[id, :ambulance] = amb

        event = gui_event("call responded", problem.emergency_calls[id, :neighborhood], i, id, amb,-1, problem.available[i] ,Dates.now())
        num_ambulances_array = Array{Integer}(undef, size(problem.available, 1))
        for i in 1:size(problem.available, 1)
            num_ambulances_array[i] = problem.available[i]
        end
        push!(ems.guiArray,event)
        push!(ems.guiArray,num_ambulances_array)
        enqueue!(ems.eventqueue, (:station2call, id, t + travel_time, amb), t + travel_time)
    #else queue it
    else
        # println(id, ": call from ", nbhd, " queued behind ", problem.wait_queue[nbhd])
        # event = gui_event("call made", nbhd, -1, id, -1,-1,-1,Dates.now())
        # push!(ems.guiArray,event)
        # problem.shortfalls = problem.shortfalls + 1
        # #push!(problem.shortfalls) # count shortfalls
        # push!(problem.wait_queue[nbhd], id) # queue the emergency call
        event = gui_event("shortfall", problem.emergency_calls[id, :neighborhood], -1, id, -1,-1,-1 ,Dates.now())
        push!(ems.guiArray, event)
        enqueue!(ems.shortqueue, (:call, id, t, problem.emergency_calls[id, :neighborhood]), t)
        println(id, ": call from ", nbhd, " is a shortfall")
    end
end

function station2call_event!(
        ems::EMSEngine,
        problem::DispatchProblem,
        dispatch::DispatchModel,
        id::Int, # the id of the emergency call
        t::Int, # the time of the emergency call
        amb::Int
    )
    arriveatscene!(dispatch, amb, t)
    # time the ambulance spends at the scene
    scene_time = ceil(Int,60*0.4*rand(problem.turnaround)) # 60sec*0.4*mean(40minutes) ~ 15minutes
    ems.eventlog[id, :scenetime] = scene_time / 60 # minutes
    @assert scene_time > 0
    
    event = gui_event("ambulance arrived", problem.emergency_calls[id, :neighborhood], -1, id, amb, t,-1 ,Dates.now())
    push!(ems.guiArray, event)
    enqueue!(ems.eventqueue, (:call2hospital, id, t + scene_time, amb), t + scene_time)
end

function call2hosp_event!(
        ems::EMSEngine,
        problem::DispatchProblem,
        dispatch::DispatchModel,
        id::Int, # the id of the emergency call
        t::Int, # the time of the emergency call
        amb::Int
    )
    h = let mintime = Inf, minindex = 0
        for h in 1:nrow(problem.hospitals)
            traveltime = problem.emergency_calls[id, Symbol("hosp$(h)_min")]
            if !isna(traveltime) && !(traveltime=="NA") && traveltime < mintime
                @assert traveltime >= 0
                minindex = h; mintime = traveltime
            end
        end
        minindex
    end
    @assert h != 0
    dispatch.hospital[amb] = ems.eventlog[id, :hospital] = h
    conveytime = 60*15 + ceil(Int, 60*problem.emergency_calls[id, Symbol("hosp$(h)_min")]) # ~20minutes
    #conveytime = ceil(conveytime * .9) #time mod
    ems.eventlog[id, :conveytime] = conveytime / 60 # minutes
    @assert conveytime >= 0 conveytime
    arriveathospital!(dispatch, amb, h, t)
    
    #         #call made / call responded / ambulance arrived / shortfall
#     event_type::String
#     #neighborhood where call is made from
#     neighborhood_id::Int
#     # station where call is made from
#     deployment_id::Int
#     # id of the event generated
#     event_id::Int
#     ambulance_id::Int
#     arrival_time::Int
#     remaining_amb::Int
#     # current timestamp of event
#     timestamp::DateTime
    event = gui_event("ambulance hospital", problem.emergency_calls[id, :neighborhood], -1, id, amb, t,-1 ,Dates.now())
    push!(ems.guiArray, event)
    
    
    enqueue!(ems.eventqueue, (:hospital2station, id, t+conveytime, amb), t+conveytime)
end

function return_event!(
        ems::EMSEngine,
        problem::DispatchProblem,
        dispatch::DispatchModel,
        id::Int,
        t::Int,
        amb::Int
    )
    stn = returning_to!(dispatch, amb, t)
    h = dispatch.hospital[amb]
    returntime = ceil(Int,60*problem.hospitals[h, Symbol("stn$(stn)_min")]) # ~ 10minutes
    #returntime = ceil(returntime * .9) #time mod
    dispatch.hospital[amb] = 0
    ems.eventlog[id, :returntime] = returntime / 60 # minutes
    @assert returntime >= 0 returntime
    t_end = t + returntime
    
#         #call made / call responded / ambulance arrived / shortfall
#     event_type::String
#     #neighborhood where call is made from
#     neighborhood_id::Int
#     # station where call is made from
#     deployment_id::Int
#     # id of the event generated
#     event_id::Int
#     ambulance_id::Int
#     arrival_time::Int
#     remaining_amb::Int
#     # current timestamp of event
#     timestamp::DateTime
    
    event = gui_event("ambulance returning", problem.emergency_calls[id, :neighborhood], -1, id, amb, t,-1 ,Dates.now())
    push!(ems.guiArray, event)
    
    
    enqueue!(ems.eventqueue, (:done, id, t_end, amb), t_end)
    # num_ambulances_array = Array{Integer}(undef, size(problem.available, 1))
    # for i in 1:size(problem.available, 1)
    #     num_ambulances_array[i] = problem.available[i]
    # end
    # push!(ems.guiArray,num_ambulances_array)
end

function done_event!(
        ems::EMSEngine,
        problem::DispatchProblem,
        dispatch::DispatchModel,
        id::Int,
        t::Int,
        amb::Int
    )
    i = dispatch.assignment[amb]
    println("Ambulance from event $id has returned to station $i")
    @assert i > 0 "$id: dispatch from $i to nbhd $nbhd" # assume valid i (enforced by <if> condition)
    update_ambulances!(dispatch, i, 1)
    push!(dispatch.ambulances[i], amb)
    dispatch.status[amb] = :available
    ems.eventlog[id, :return_to] = id
    @assert problem.available[i] > 0
    
#         #call made / call responded / ambulance arrived / shortfall
#     event_type::String
#     #neighborhood where call is made from
#     neighborhood_id::Int
#     # station where call is made from
#     deployment_id::Int
#     # id of the event generated
#     event_id::Int
#     ambulance_id::Int
#     arrival_time::Int
#     remaining_amb::Int
#     # current timestamp of event
#     timestamp::DateTime
    
    event = gui_event("ambulance returned", problem.emergency_calls[id, :neighborhood], -1, id, amb, t,-1 ,Dates.now())
    push!(ems.guiArray, event)
    
    while(~isempty(ems.shortqueue))
        # record time difference
        (event, short_id, short_t, value) = dequeue!(ems.shortqueue)
        enqueue!(ems.eventqueue, (event, short_id, t , value), short_t)
    end
end


  Resolving package versions...
No Changes to `C:\Users\Owner\Documents\Austin\AmbulanceDeployment\AmbulanceDeployment.jl-legacy\Project.toml`
No Changes to `C:\Users\Owner\Documents\Austin\AmbulanceDeployment\AmbulanceDeployment.jl-legacy\Manifest.toml`


done_event! (generic function with 1 method)

In [63]:
"""call simulation"""
#@time df, guiArray = simulate_events!(problem, dispatch)
df, guiArray = simulate_events!(problem, dispatch)

calling event id: 1 time: 38 value: 83
Ambulance has arrived on the scene for event 1 at time 192
calling event id: 2 time: 575 value: 38
calling event id: 3 time: 673 value: 129
calling event id: 4 time: 691 value: 107
Ambulance for event 1 has arrived at the hospital at time 875
Ambulance has arrived on the scene for event 2 at time 1104
Ambulance has arrived on the scene for event 4 at time 1156
calling event id: 5 time: 1166 value: 113
calling event id: 6 time: 1187 value: 51
Ambulance has arrived on the scene for event 3 at time 1261
Ambulance has arrived on the scene for event 6 at time 1341
calling event id: 7 time: 1450 value: 76
calling event id: 8 time: 1578 value: 67
Ambulance has arrived on the scene for event 5 at time 1606
calling event id: 9 time: 1640 value: 52
Ambulance has arrived on the scene for event 8 at time 1732
calling event id: 10 time: 1838 value: 67
Ambulance for event 4 has arrived at the hospital at time 1892
calling event id: 11 time: 1899 value: 67
calli

Ambulance has arrived on the scene for event 34 at time 9823
Ambulance for event 33 has arrived at the hospital at time 9874
Ambulance has arrived on the scene for event 36 at time 10006
calling event id: 37 time: 10091 value: 98
Ambulance for event 29 is returning to the station at 10125
Ambulance from event 21 has returned to station 43
Ambulance for event 27 is returning to the station at 10317
Ambulance for event 32 has arrived at the hospital at time 10404
Ambulance for event 34 has arrived at the hospital at time 10494
Ambulance from event 27 has returned to station 17
Ambulance from event 29 has returned to station 3
Ambulance for event 35 has arrived at the hospital at time 10543
Ambulance has arrived on the scene for event 37 at time 10553
calling event id: 38 time: 10595 value: 98
Ambulance for event 30 is returning to the station at 10657
Ambulance for event 31 is returning to the station at 10686
calling event id: 39 time: 10727 value: 52
calling event id: 40 time: 10770 va

Ambulance from event 55 has returned to station 17
calling event id: 63 time: 19271 value: 98
Ambulance for event 60 is returning to the station at 19386
calling event id: 64 time: 19691 value: 51
Ambulance has arrived on the scene for event 63 at time 19733
Ambulance has arrived on the scene for event 64 at time 19845
Ambulance for event 62 has arrived at the hospital at time 19867
calling event id: 65 time: 19963 value: 98
Ambulance for event 59 is returning to the station at 19969
Ambulance from event 60 has returned to station 1
Ambulance for event 61 has arrived at the hospital at time 20238
calling event id: 66 time: 20471 value: 113
Ambulance from event 59 has returned to station 14
Ambulance has arrived on the scene for event 65 at time 20639
calling event id: 67 time: 20814 value: 36
Ambulance for event 64 has arrived at the hospital at time 20831
Ambulance for event 63 has arrived at the hospital at time 20909
Ambulance has arrived on the scene for event 66 at time 20911
call

Ambulance for event 85 is returning to the station at 31244
Ambulance for event 89 is returning to the station at 31826
Ambulance for event 92 is returning to the station at 32057
Ambulance from event 87 has returned to station 22
Ambulance from event 85 has returned to station 22
Ambulance from event 89 has returned to station 25
calling event id: 93 time: 32356 value: 66
Ambulance for event 91 is returning to the station at 32422
Ambulance has arrived on the scene for event 93 at time 32510
Ambulance from event 91 has returned to station 8
Ambulance from event 92 has returned to station 14
calling event id: 94 time: 32957 value: 66
Ambulance for event 90 is returning to the station at 33079
Ambulance has arrived on the scene for event 94 at time 33111
Ambulance from event 90 has returned to station 39
Ambulance for event 93 has arrived at the hospital at time 33418
calling event id: 95 time: 33904 value: 99
calling event id: 96 time: 33986 value: 142
Ambulance has arrived on the scen

Ambulance from event 113 has returned to station 14
calling event id: 126 time: 45420 value: 143
Ambulance for event 116 has arrived at the hospital at time 45430
Ambulance for event 117 has arrived at the hospital at time 45536
Ambulance has arrived on the scene for event 124 at time 45708
calling event id: 127 time: 45780 value: 97
Ambulance has arrived on the scene for event 126 at time 45830
Ambulance for event 122 has arrived at the hospital at time 45880
Ambulance for event 120 has arrived at the hospital at time 45904
Ambulance for event 115 is returning to the station at 45926
Ambulance has arrived on the scene for event 127 at time 46008
Ambulance from event 115 has returned to station 8
calling event id: 128 time: 46110 value: 82
Ambulance has arrived on the scene for event 125 at time 46143
Ambulance for event 121 has arrived at the hospital at time 46232
Ambulance for event 123 has arrived at the hospital at time 46491
calling event id: 129 time: 46504 value: 141
calling ev

calling event id: 154 time: 53031 value: 83
Ambulance for event 140 is returning to the station at 53049
Ambulance from event 133 has returned to station 30
Ambulance for event 148 has arrived at the hospital at time 53147
calling event id: 155 time: 53166 value: 128
Ambulance from event 141 has returned to station 5
calling event id: 156 time: 53203 value: 98
calling event id: 157 time: 53232 value: 65
Ambulance for event 144 is returning to the station at 53311
Ambulance has arrived on the scene for event 155 at time 53320
Ambulance for event 150 has arrived at the hospital at time 53361
Ambulance from event 136 has returned to station 2
calling event id: 158 time: 53375 value: 128
Ambulance has arrived on the scene for event 157 at time 53386
Ambulance from event 140 has returned to station 3
Ambulance for event 142 is returning to the station at 53421
Ambulance has arrived on the scene for event 153 at time 53488
Ambulance from event 139 has returned to station 19
calling event id:

calling event id: 181 time: 58743 value: 129
181: call from 129 is a shortfall
Ambulance for event 173 has arrived at the hospital at time 58832
Ambulance for event 176 has arrived at the hospital at time 58847
Ambulance has arrived on the scene for event 180 at time 58962
Ambulance has arrived on the scene for event 183 at time 59053
Ambulance for event 175 has arrived at the hospital at time 59101
calling event id: 185 time: 59176 value: 143
calling event id: 186 time: 59346 value: 97
Ambulance has arrived on the scene for event 184 at time 59383
Ambulance for event 172 is returning to the station at 59485
Ambulance for event 182 has arrived at the hospital at time 59540
Ambulance has arrived on the scene for event 185 at time 59624
Ambulance from event 168 has returned to station 44
calling event id: 181 time: 59641 value: 129
181: call from 129 is a shortfall
Ambulance has arrived on the scene for event 186 at time 59645
Ambulance for event 179 has arrived at the hospital at time 5

189: call from 125 is a shortfall
Ambulance for event 203 has arrived at the hospital at time 63716
calling event id: 207 time: 63762 value: 66
Ambulance has arrived on the scene for event 206 at time 63881
Ambulance for event 195 is returning to the station at 63915
Ambulance has arrived on the scene for event 207 at time 63916
calling event id: 208 time: 63937 value: 51
Ambulance for event 204 has arrived at the hospital at time 64005
calling event id: 209 time: 64027 value: 35
Ambulance from event 174 has returned to station 7
calling event id: 189 time: 64067 value: 125
This is a shortfallAmbulance from event 181 has returned to station 14
Ambulance has arrived on the scene for event 209 at time 64181
Ambulance has arrived on the scene for event 202 at time 64206
Ambulance for event 198 is returning to the station at 64211
Ambulance from event 190 has returned to station 8
Ambulance for event 194 is returning to the station at 64252
Ambulance for event 200 is returning to the stati

Ambulance for event 226 has arrived at the hospital at time 70002
Ambulance for event 218 is returning to the station at 70024
Ambulance for event 222 is returning to the station at 70030
calling event id: 239 time: 70076 value: 98
Ambulance for event 228 has arrived at the hospital at time 70077
Ambulance has arrived on the scene for event 238 at time 70102
calling event id: 240 time: 70103 value: 124
Ambulance from event 220 has returned to station 11
Ambulance has arrived on the scene for event 227 at time 70192
calling event id: 241 time: 70238 value: 116
Ambulance has arrived on the scene for event 240 at time 70257
calling event id: 242 time: 70267 value: 116
242: call from 116 is a shortfall
Ambulance has arrived on the scene for event 237 at time 70278
Ambulance has arrived on the scene for event 236 at time 70362
Ambulance from event 222 has returned to station 3
calling event id: 242 time: 70441 value: 116
242: call from 116 is a shortfall
Ambulance has arrived on the scene f

265: call from 82 is a shortfall
Ambulance has arrived on the scene for event 257 at time 74734
Ambulance has arrived on the scene for event 260 at time 74747
Ambulance from event 229 has returned to station 33
calling event id: 265 time: 74846 value: 82
265: call from 82 is a shortfall
Ambulance has arrived on the scene for event 262 at time 74889
Ambulance for event 254 has arrived at the hospital at time 74907
Ambulance has arrived on the scene for event 263 at time 74935
calling event id: 266 time: 74941 value: 98
266: call from 98 is a shortfall
Ambulance from event 245 has returned to station 25
calling event id: 265 time: 74977 value: 82
265: call from 82 is a shortfall
calling event id: 266 time: 74977 value: 98
266: call from 98 is a shortfall
calling event id: 267 time: 75015 value: 68
267: call from 68 is a shortfall
Ambulance for event 253 has arrived at the hospital at time 75028
calling event id: 268 time: 75093 value: 67
Ambulance for event 242 is returning to the statio

calling event id: 284 time: 78493 value: 23
calling event id: 285 time: 78502 value: 113
Ambulance from event 265 has returned to station 8
Ambulance from event 259 has returned to station 44
Ambulance has arrived on the scene for event 283 at time 78572
Ambulance for event 270 is returning to the station at 78721
Ambulance for event 273 is returning to the station at 78788
Ambulance for event 275 is returning to the station at 78821
Ambulance for event 277 has arrived at the hospital at time 78862
Ambulance from event 268 has returned to station 18
Ambulance for event 279 has arrived at the hospital at time 78875
Ambulance has arrived on the scene for event 285 at time 78942
Ambulance for event 278 has arrived at the hospital at time 78944
Ambulance for event 274 is returning to the station at 78948
Ambulance has arrived on the scene for event 284 at time 78955
Ambulance for event 272 is returning to the station at 78965
Ambulance for event 281 has arrived at the hospital at time 7896

Ambulance from event 301 has returned to station 15
Ambulance for event 303 is returning to the station at 85588
Ambulance for event 294 is returning to the station at 85622
Ambulance has arrived on the scene for event 307 at time 85686
Ambulance from event 294 has returned to station 34
Ambulance from event 290 has returned to station 33
calling event id: 311 time: 85942 value: 98
Ambulance has arrived on the scene for event 309 at time 85963
calling event id: 312 time: 86001 value: 98
calling event id: 313 time: 86125 value: 98
Ambulance has arrived on the scene for event 310 at time 86150
Ambulance for event 302 is returning to the station at 86262
Ambulance for event 306 is returning to the station at 86268
Ambulance for event 304 is returning to the station at 86277
Ambulance for event 308 has arrived at the hospital at time 86303
Ambulance for event 307 has arrived at the hospital at time 86351
Ambulance has arrived on the scene for event 311 at time 86404
calling event id: 314 t

calling event id: 338 time: 92208 value: 52
Ambulance for event 328 has arrived at the hospital at time 92268
Ambulance for event 323 is returning to the station at 92364
Ambulance has arrived on the scene for event 320 at time 92365
Ambulance from event 325 has returned to station 26
calling event id: 339 time: 92457 value: 82
Ambulance has arrived on the scene for event 337 at time 92502
Ambulance has arrived on the scene for event 336 at time 92586
calling event id: 340 time: 92589 value: 65
Ambulance has arrived on the scene for event 339 at time 92611
Ambulance for event 329 is returning to the station at 92662
Ambulance has arrived on the scene for event 334 at time 92663
Ambulance for event 333 has arrived at the hospital at time 92687
Ambulance has arrived on the scene for event 340 at time 92743
Ambulance for event 331 has arrived at the hospital at time 92754
Ambulance has arrived on the scene for event 338 at time 92948
Ambulance for event 320 has arrived at the hospital at 

Ambulance from event 351 has returned to station 8
calling event id: 362 time: 100202 value: 67
Ambulance from event 352 has returned to station 3
Ambulance from event 356 has returned to station 5
Ambulance from event 357 has returned to station 17
Ambulance has arrived on the scene for event 360 at time 100349
Ambulance has arrived on the scene for event 362 at time 100356
Ambulance from event 354 has returned to station 14
Ambulance for event 355 is returning to the station at 100520
Ambulance has arrived on the scene for event 361 at time 100719
calling event id: 363 time: 100728 value: 67
Ambulance for event 359 has arrived at the hospital at time 100767
Ambulance has arrived on the scene for event 363 at time 100882
Ambulance from event 350 has returned to station 43
Ambulance for event 361 has arrived at the hospital at time 101095
Ambulance from event 355 has returned to station 25
Ambulance for event 360 has arrived at the hospital at time 101284
Ambulance for event 362 has ar

Ambulance has arrived on the scene for event 391 at time 111830
Ambulance for event 383 is returning to the station at 111991
Ambulance has arrived on the scene for event 389 at time 112031
Ambulance for event 387 has arrived at the hospital at time 112094
Ambulance from event 383 has returned to station 8
Ambulance for event 384 is returning to the station at 112350
Ambulance for event 385 is returning to the station at 112511
Ambulance for event 390 has arrived at the hospital at time 112582
calling event id: 392 time: 112670 value: 113
Ambulance from event 384 has returned to station 3
Ambulance for event 391 has arrived at the hospital at time 112831
Ambulance from event 385 has returned to station 4
Ambulance for event 389 has arrived at the hospital at time 113105
Ambulance has arrived on the scene for event 392 at time 113110
Ambulance for event 386 has arrived at the hospital at time 113312
Ambulance for event 387 is returning to the station at 113580
calling event id: 393 time

Ambulance for event 407 is returning to the station at 126954
Ambulance for event 416 has arrived at the hospital at time 127003
Ambulance has arrived on the scene for event 418 at time 127033
Ambulance for event 410 is returning to the station at 127070
Ambulance for event 411 is returning to the station at 127194
Ambulance from event 410 has returned to station 8
Ambulance from event 409 has returned to station 16
Ambulance from event 411 has returned to station 17
Ambulance from event 407 has returned to station 14
Ambulance for event 417 has arrived at the hospital at time 127718
Ambulance for event 412 is returning to the station at 127783
calling event id: 419 time: 127930 value: 98
calling event id: 420 time: 127932 value: 81
Ambulance for event 414 is returning to the station at 127950
Ambulance for event 418 has arrived at the hospital at time 127955
Ambulance for event 415 is returning to the station at 128003
calling event id: 421 time: 128056 value: 67
Ambulance has arrived

calling event id: 449 time: 135672 value: 65
calling event id: 450 time: 135673 value: 65
Ambulance has arrived on the scene for event 446 at time 135712
calling event id: 451 time: 135755 value: 69
calling event id: 452 time: 135813 value: 82
Ambulance has arrived on the scene for event 449 at time 135826
Ambulance has arrived on the scene for event 450 at time 135827
calling event id: 453 time: 135843 value: 52
Ambulance for event 442 has arrived at the hospital at time 135872
calling event id: 454 time: 135893 value: 82
Ambulance for event 439 is returning to the station at 135927
Ambulance has arrived on the scene for event 452 at time 135967
Ambulance for event 444 has arrived at the hospital at time 135968
calling event id: 455 time: 135989 value: 65
Ambulance for event 443 has arrived at the hospital at time 136006
Ambulance has arrived on the scene for event 447 at time 136049
calling event id: 456 time: 136071 value: 67
Ambulance has arrived on the scene for event 451 at time 

Ambulance for event 469 has arrived at the hospital at time 139907
Ambulance for event 470 has arrived at the hospital at time 139912
Ambulance for event 472 has arrived at the hospital at time 139919
Ambulance from event 462 has returned to station 14
Ambulance for event 473 has arrived at the hospital at time 140062
Ambulance from event 457 has returned to station 28
Ambulance has arrived on the scene for event 476 at time 140097
Ambulance for event 471 has arrived at the hospital at time 140131
Ambulance for event 463 is returning to the station at 140143
Ambulance from event 464 has returned to station 4
Ambulance has arrived on the scene for event 477 at time 140302
calling event id: 478 time: 140344 value: 112
Ambulance for event 474 has arrived at the hospital at time 140418
Ambulance for event 466 is returning to the station at 140426
Ambulance from event 466 has returned to station 17
Ambulance for event 475 has arrived at the hospital at time 140666
calling event id: 479 time

Ambulance has arrived on the scene for event 501 at time 146828
Ambulance for event 500 has arrived at the hospital at time 146858
Ambulance from event 495 has returned to station 3
Ambulance for event 492 is returning to the station at 146919
calling event id: 504 time: 146982 value: 113
calling event id: 505 time: 146985 value: 22
Ambulance for event 499 has arrived at the hospital at time 146986
Ambulance from event 488 has returned to station 2
calling event id: 506 time: 147068 value: 144
Ambulance from event 491 has returned to station 15
Ambulance for event 494 is returning to the station at 147116
Ambulance has arrived on the scene for event 505 at time 147139
Ambulance for event 497 is returning to the station at 147181
calling event id: 507 time: 147259 value: 98
Ambulance has arrived on the scene for event 503 at time 147270
Ambulance has arrived on the scene for event 504 at time 147422
Ambulance for event 496 is returning to the station at 147486
calling event id: 508 time

528: call from 52 is a shortfall
calling event id: 531 time: 150780 value: 130
531: call from 130 is a shortfall
Ambulance from event 512 has returned to station 17
calling event id: 528 time: 150781 value: 52
528: call from 52 is a shortfall
calling event id: 531 time: 150781 value: 130
531: call from 130 is a shortfall
calling event id: 533 time: 150792 value: 113
Ambulance for event 508 is returning to the station at 150815
Ambulance from event 515 has returned to station 18
calling event id: 528 time: 150815 value: 52
528: call from 52 is a shortfall
calling event id: 531 time: 150815 value: 130
531: call from 130 is a shortfall
Ambulance for event 519 is returning to the station at 150848
Ambulance for event 527 has arrived at the hospital at time 150858
Ambulance from event 506 has returned to station 25
calling event id: 528 time: 150887 value: 52
528: call from 52 is a shortfall
calling event id: 531 time: 150887 value: 130
531: call from 130 is a shortfall
Ambulance has arrive

Ambulance for event 533 is returning to the station at 153517
Ambulance from event 522 has returned to station 22
calling event id: 531 time: 153600 value: 130
531: call from 130 is a shortfall
calling event id: 539 time: 153600 value: 129
539: call from 129 is a shortfall
calling event id: 540 time: 153600 value: 129
540: call from 129 is a shortfall
calling event id: 541 time: 153600 value: 129
541: call from 129 is a shortfall
Ambulance from event 533 has returned to station 17
calling event id: 531 time: 153721 value: 130
531: call from 130 is a shortfall
calling event id: 541 time: 153721 value: 129
541: call from 129 is a shortfall
calling event id: 540 time: 153721 value: 129
540: call from 129 is a shortfall
calling event id: 539 time: 153721 value: 129
539: call from 129 is a shortfall
Ambulance for event 534 is returning to the station at 153755
calling event id: 543 time: 153802 value: 98
Ambulance for event 538 has arrived at the hospital at time 153982
Ambulance for event 

Ambulance for event 548 is returning to the station at 157893
Ambulance from event 539 has returned to station 25
calling event id: 541 time: 157919 value: 129
This is a shortfallcalling event id: 547 time: 157919 value: 129
547: call from 129 is a shortfall
calling event id: 559 time: 157933 value: 51
Ambulance for event 540 is returning to the station at 157954
Ambulance for event 555 has arrived at the hospital at time 158024
calling event id: 560 time: 158033 value: 83
calling event id: 561 time: 158054 value: 82
Ambulance has arrived on the scene for event 559 at time 158087
Ambulance for event 557 has arrived at the hospital at time 158094
Ambulance has arrived on the scene for event 560 at time 158187
calling event id: 562 time: 158190 value: 82
Ambulance for event 552 is returning to the station at 158208
Ambulance has arrived on the scene for event 561 at time 158208
calling event id: 563 time: 158225 value: 113
Ambulance from event 549 has returned to station 15
calling event

Ambulance from event 573 has returned to station 19
Ambulance for event 576 is returning to the station at 164537
Ambulance for event 579 has arrived at the hospital at time 164645
Ambulance for event 575 is returning to the station at 164772
Ambulance for event 583 has arrived at the hospital at time 164852
calling event id: 584 time: 164896 value: 36
Ambulance from event 575 has returned to station 17
calling event id: 585 time: 165068 value: 67
Ambulance for event 577 is returning to the station at 165161
Ambulance has arrived on the scene for event 585 at time 165222
Ambulance for event 581 has arrived at the hospital at time 165346
Ambulance from event 577 has returned to station 17
Ambulance from event 576 has returned to station 2
Ambulance for event 582 has arrived at the hospital at time 165430
Ambulance for event 580 has arrived at the hospital at time 165455
Ambulance has arrived on the scene for event 584 at time 165497
calling event id: 586 time: 165630 value: 128
Ambulanc

Ambulance has arrived on the scene for event 611 at time 173093
Ambulance has arrived on the scene for event 612 at time 173107
Ambulance for event 604 is returning to the station at 173126
calling event id: 613 time: 173165 value: 68
calling event id: 614 time: 173168 value: 142
Ambulance for event 603 is returning to the station at 173181
Ambulance from event 605 has returned to station 34
calling event id: 615 time: 173235 value: 98
Ambulance for event 608 is returning to the station at 173261
Ambulance from event 602 has returned to station 14
Ambulance has arrived on the scene for event 613 at time 173319
calling event id: 616 time: 173350 value: 67
Ambulance from event 603 has returned to station 17
Ambulance for event 599 is returning to the station at 173427
Ambulance for event 606 is returning to the station at 173492
Ambulance from event 608 has returned to station 34
Ambulance has arrived on the scene for event 616 at time 173504
Ambulance has arrived on the scene for event 

calling event id: 647 time: 179536 value: 98
calling event id: 648 time: 179582 value: 56
calling event id: 649 time: 179594 value: 51
Ambulance for event 628 is returning to the station at 179611
Ambulance from event 617 has returned to station 32
calling event id: 650 time: 179659 value: 67
Ambulance has arrived on the scene for event 645 at time 179681
Ambulance from event 626 has returned to station 9
Ambulance has arrived on the scene for event 649 at time 179748
Ambulance from event 629 has returned to station 28
Ambulance for event 640 has arrived at the hospital at time 179831
Ambulance for event 633 is returning to the station at 179865
Ambulance has arrived on the scene for event 646 at time 179890
Ambulance for event 630 is returning to the station at 179925
Ambulance for event 644 has arrived at the hospital at time 179935
Ambulance for event 642 has arrived at the hospital at time 180054
Ambulance from event 628 has returned to station 11
Ambulance has arrived on the scene

667: call from 84 is a shortfall
Ambulance for event 666 has arrived at the hospital at time 184759
calling event id: 672 time: 185022 value: 52
calling event id: 673 time: 185072 value: 67
Ambulance for event 668 has arrived at the hospital at time 185171
Ambulance has arrived on the scene for event 673 at time 185226
Ambulance for event 662 is returning to the station at 185257
Ambulance for event 669 has arrived at the hospital at time 185335
calling event id: 674 time: 185342 value: 103
Ambulance has arrived on the scene for event 671 at time 185350
Ambulance from event 660 has returned to station 17
calling event id: 667 time: 185492 value: 84
667: call from 84 is a shortfall
Ambulance for event 670 has arrived at the hospital at time 185547
Ambulance has arrived on the scene for event 672 at time 185557
Ambulance for event 663 has arrived at the hospital at time 185566
Ambulance for event 671 has arrived at the hospital at time 185705
Ambulance from event 662 has returned to stat

calling event id: 693 time: 196136 value: 98
Ambulance has arrived on the scene for event 692 at time 196197
calling event id: 694 time: 196228 value: 51
Ambulance from event 691 has returned to station 3
Ambulance has arrived on the scene for event 694 at time 196382
Ambulance has arrived on the scene for event 693 at time 196596
Ambulance for event 692 has arrived at the hospital at time 196947
Ambulance for event 690 is returning to the station at 197426
Ambulance for event 694 has arrived at the hospital at time 197470
Ambulance for event 693 has arrived at the hospital at time 197473
calling event id: 695 time: 197712 value: 113
Ambulance from event 690 has returned to station 4
calling event id: 696 time: 198011 value: 67
calling event id: 697 time: 198122 value: 113
Ambulance has arrived on the scene for event 695 at time 198152
Ambulance has arrived on the scene for event 696 at time 198165
calling event id: 698 time: 198430 value: 113
Ambulance for event 692 is returning to th

Ambulance for event 719 has arrived at the hospital at time 212998
calling event id: 721 time: 213169 value: 82
Ambulance has arrived on the scene for event 721 at time 213323
Ambulance for event 720 has arrived at the hospital at time 213455
calling event id: 722 time: 213675 value: 128
Ambulance has arrived on the scene for event 722 at time 213829
calling event id: 723 time: 213866 value: 67
Ambulance has arrived on the scene for event 723 at time 214020
Ambulance for event 721 has arrived at the hospital at time 214131
calling event id: 724 time: 214214 value: 54
Ambulance has arrived on the scene for event 724 at time 214368
calling event id: 725 time: 214471 value: 66
Ambulance for event 719 is returning to the station at 214484
Ambulance for event 722 has arrived at the hospital at time 214545
Ambulance has arrived on the scene for event 725 at time 214625
calling event id: 726 time: 214820 value: 82
Ambulance for event 723 has arrived at the hospital at time 214862
calling even

Ambulance has arrived on the scene for event 757 at time 222669
Ambulance has arrived on the scene for event 756 at time 222719
Ambulance for event 747 is returning to the station at 222953
calling event id: 760 time: 222958 value: 113
Ambulance for event 750 has arrived at the hospital at time 222972
Ambulance for event 739 is returning to the station at 223077
Ambulance for event 746 is returning to the station at 223111
Ambulance has arrived on the scene for event 760 at time 223112
Ambulance for event 749 is returning to the station at 223127
Ambulance has arrived on the scene for event 758 at time 223133
Ambulance for event 745 is returning to the station at 223211
Ambulance from event 749 has returned to station 8
calling event id: 752 time: 223279 value: 84
752: call from 84 is a shortfall
calling event id: 759 time: 223279 value: 84
759: call from 84 is a shortfall
Ambulance for event 754 has arrived at the hospital at time 223292
Ambulance for event 742 is returning to the sta

calling event id: 765 time: 227391 value: 84
This is a shortfallAmbulance for event 768 is returning to the station at 227471
Ambulance for event 773 has arrived at the hospital at time 227569
Ambulance for event 770 has arrived at the hospital at time 227666
calling event id: 777 time: 227701 value: 67
Ambulance has arrived on the scene for event 776 at time 227714
Ambulance for event 766 is returning to the station at 227817
calling event id: 778 time: 227837 value: 50
Ambulance for event 759 is returning to the station at 227858
Ambulance has arrived on the scene for event 765 at time 227915
Ambulance from event 768 has returned to station 25
Ambulance for event 774 has arrived at the hospital at time 228060
Ambulance has arrived on the scene for event 777 at time 228132
calling event id: 779 time: 228155 value: 67
calling event id: 780 time: 228174 value: 82
calling event id: 781 time: 228246 value: 51
Ambulance for event 775 has arrived at the hospital at time 228294
Ambulance for

Ambulance for event 789 is returning to the station at 233388
Ambulance for event 799 has arrived at the hospital at time 233444
Ambulance has arrived on the scene for event 801 at time 233463
Ambulance for event 800 has arrived at the hospital at time 233463
Ambulance for event 797 has arrived at the hospital at time 233532
calling event id: 804 time: 233570 value: 38
Ambulance from event 789 has returned to station 17
calling event id: 798 time: 233592 value: 84
798: call from 84 is a shortfall
Ambulance has arrived on the scene for event 802 at time 233595
Ambulance for event 793 is returning to the station at 233680
Ambulance from event 790 has returned to station 9
calling event id: 798 time: 233731 value: 84
798: call from 84 is a shortfall
calling event id: 805 time: 233839 value: 68
Ambulance from event 788 has returned to station 14
calling event id: 798 time: 233847 value: 84
798: call from 84 is a shortfall
Ambulance has arrived on the scene for event 803 at time 233851
Ambu

Ambulance for event 823 has arrived at the hospital at time 239369
Ambulance for event 825 has arrived at the hospital at time 239447
Ambulance from event 817 has returned to station 8
Ambulance for event 822 has arrived at the hospital at time 239495
Ambulance for event 820 has arrived at the hospital at time 239532
Ambulance for event 826 has arrived at the hospital at time 239661
Ambulance has arrived on the scene for event 831 at time 239688
Ambulance for event 824 has arrived at the hospital at time 239699
Ambulance for event 821 is returning to the station at 239700
calling event id: 833 time: 239700 value: 66
Ambulance for event 830 has arrived at the hospital at time 239731
Ambulance has arrived on the scene for event 832 at time 239800
Ambulance for event 828 has arrived at the hospital at time 239939
Ambulance from event 821 has returned to station 5
calling event id: 834 time: 240077 value: 82
Ambulance for event 819 is returning to the station at 240105
calling event id: 83

Ambulance from event 835 has returned to station 16
calling event id: 858 time: 244231 value: 84
858: call from 84 is a shortfall
Ambulance for event 850 has arrived at the hospital at time 244286
calling event id: 861 time: 244381 value: 66
Ambulance for event 849 has arrived at the hospital at time 244390
Ambulance for event 854 has arrived at the hospital at time 244399
calling event id: 862 time: 244416 value: 141
Ambulance for event 852 has arrived at the hospital at time 244423
Ambulance for event 848 has arrived at the hospital at time 244442
Ambulance for event 855 has arrived at the hospital at time 244444
Ambulance for event 856 has arrived at the hospital at time 244460
Ambulance has arrived on the scene for event 859 at time 244506
Ambulance from event 836 has returned to station 2
calling event id: 858 time: 244535 value: 84
858: call from 84 is a shortfall
Ambulance has arrived on the scene for event 862 at time 244570
Ambulance for event 851 has arrived at the hospital a

calling event id: 888 time: 249530 value: 98
calling event id: 889 time: 249560 value: 37
889: call from 37 is a shortfall
Ambulance from event 863 has returned to station 31
calling event id: 884 time: 249567 value: 52
884: call from 52 is a shortfall
calling event id: 889 time: 249567 value: 37
This is a shortfallAmbulance for event 876 has arrived at the hospital at time 249578
Ambulance for event 874 has arrived at the hospital at time 249636
Ambulance has arrived on the scene for event 886 at time 249774
Ambulance for event 883 has arrived at the hospital at time 249817
Ambulance has arrived on the scene for event 887 at time 249891
Ambulance for event 881 has arrived at the hospital at time 249913
Ambulance from event 870 has returned to station 3
calling event id: 884 time: 249979 value: 52
884: call from 52 is a shortfall
Ambulance has arrived on the scene for event 888 at time 249992
Ambulance from event 858 has returned to station 44
calling event id: 884 time: 250030 value: 

Ambulance from event 899 has returned to station 8
Ambulance for event 903 is returning to the station at 255535
Ambulance has arrived on the scene for event 910 at time 255553
Ambulance has arrived on the scene for event 912 at time 255563
Ambulance for event 904 is returning to the station at 255610
Ambulance for event 901 is returning to the station at 255763
Ambulance for event 896 is returning to the station at 255776
Ambulance from event 894 has returned to station 4
Ambulance has arrived on the scene for event 911 at time 255862
Ambulance from event 904 has returned to station 5
Ambulance from event 896 has returned to station 8
Ambulance for event 898 is returning to the station at 255944
Ambulance from event 903 has returned to station 11
Ambulance from event 897 has returned to station 22
Ambulance for event 910 has arrived at the hospital at time 256164
Ambulance for event 906 is returning to the station at 256199
calling event id: 913 time: 256204 value: 112
calling event i

Ambulance from event 928 has returned to station 5
calling event id: 937 time: 261778 value: 129
937: call from 129 is a shortfall
Ambulance has arrived on the scene for event 939 at time 261804
Ambulance for event 923 is returning to the station at 261811
calling event id: 941 time: 261834 value: 99
941: call from 99 is a shortfall
Ambulance from event 926 has returned to station 9
calling event id: 937 time: 262004 value: 129
937: call from 129 is a shortfall
calling event id: 941 time: 262004 value: 99
941: call from 99 is a shortfall
calling event id: 942 time: 262019 value: 51
Ambulance from event 920 has returned to station 19
calling event id: 937 time: 262028 value: 129
937: call from 129 is a shortfall
calling event id: 941 time: 262028 value: 99
941: call from 99 is a shortfall
Ambulance for event 925 is returning to the station at 262038
Ambulance for event 927 is returning to the station at 262119
Ambulance has arrived on the scene for event 942 at time 262173
Ambulance has

Ambulance from event 947 has returned to station 17
calling event id: 959 time: 265917 value: 82
Ambulance for event 949 is returning to the station at 266007
Ambulance for event 950 is returning to the station at 266046
calling event id: 960 time: 266062 value: 68
Ambulance for event 948 is returning to the station at 266124
Ambulance for event 953 is returning to the station at 266152
Ambulance for event 955 has arrived at the hospital at time 266157
Ambulance has arrived on the scene for event 960 at time 266216
Ambulance for event 951 is returning to the station at 266304
Ambulance for event 941 is returning to the station at 266323
Ambulance for event 937 is returning to the station at 266373
Ambulance has arrived on the scene for event 959 at time 266391
Ambulance from event 950 has returned to station 17
calling event id: 961 time: 266455 value: 142
Ambulance from event 951 has returned to station 8
Ambulance from event 948 has returned to station 3
calling event id: 962 time: 2

Ambulance from event 975 has returned to station 9
Ambulance for event 982 has arrived at the hospital at time 274924
Ambulance for event 976 is returning to the station at 275174
Ambulance has arrived on the scene for event 983 at time 275186
calling event id: 984 time: 275597 value: 98
calling event id: 985 time: 275709 value: 51
calling event id: 986 time: 275745 value: 98
Ambulance from event 976 has returned to station 9
Ambulance for event 983 has arrived at the hospital at time 275847
Ambulance has arrived on the scene for event 985 at time 275863
Ambulance for event 978 is returning to the station at 275865
Ambulance for event 980 is returning to the station at 275874
Ambulance from event 978 has returned to station 8
Ambulance has arrived on the scene for event 984 at time 276059
Ambulance has arrived on the scene for event 986 at time 276207
Ambulance for event 977 is returning to the station at 276224
Ambulance for event 979 is returning to the station at 276293
Ambulance fr

(1000×11 DataFrame. Omitted printing of 6 columns
│ Row  │ id    │ dispatch_from │ waittime │ responsetime │ scenetime │
│      │ Int64 │ Int64         │ Float64  │ Float64      │ Float64   │
├──────┼───────┼───────────────┼──────────┼──────────────┼───────────┤
│ 1    │ 1     │ 1             │ 0.0      │ 2.56667      │ 11.3833   │
│ 2    │ 2     │ 43            │ 0.0      │ 8.81667      │ 13.35     │
│ 3    │ 3     │ 14            │ 0.0      │ 9.8          │ 13.0      │
│ 4    │ 4     │ 34            │ 0.0      │ 7.75         │ 12.2667   │
│ 5    │ 5     │ 17            │ 0.0      │ 7.33333      │ 11.25     │
│ 6    │ 6     │ 22            │ 0.0      │ 2.56667      │ 16.25     │
│ 7    │ 7     │ 39            │ 0.0      │ 22.1167      │ 12.1333   │
│ 8    │ 8     │ 8             │ 0.0      │ 2.56667      │ 19.8333   │
│ 9    │ 9     │ 8             │ 0.0      │ 8.91667      │ 12.75     │
│ 10   │ 10    │ 3             │ 0.0      │ 7.18333      │ 9.08333   │
⋮
│ 990  │ 990   │ 44      

In [ ]:
"""batch call simulation"""
# batches = 12
# test_calls = CSV.File("../test/austin-data/austin_test_calls_update1.csv")|> DataFrame 
# #update 1: remove negative indices
# using DelimitedFiles

# for i = 1:12
#     """ simulation parameters """
#     p = DeploymentProblem(
#               hourly_calls,
#               adjacent_nbhd,
#               coverage,
#               namb = namb,
#               train_filter = (hourly_calls[!,:year] .== 2019) .* (hourly_calls[!,:month] .<= 3)
#               )
#     test_calls = CSV.File("../test/austin-data/austin_test_calls_update1.csv")|> DataFrame 
#     #update 1: remove negative indices
#     start_ind = (i-1)*1000+1
#     end_ind = (i)*1000
#     println("running $namb ambuances & $start_ind to $end_ind calls")
#     test_calls_batch = test_calls[start_ind:end_ind,:]

#     # distribution of ambulances across all of the stations
#     x = amb_deployment[model_dict[model_name]][namb]

#     #establishes the dispatch problem
#     problem = DispatchProblem(test_calls_batch, hospitals, stations, p.coverage, x, turnaround=turnaround)

#     #establishes the dispatch model
#     dispatch = ClosestDispatch(p, problem)

#     #@time df, guiArray = simulate_events!(problem, dispatch)
#     df, guiArray = simulate_events!(problem, dispatch)
#     print(mean(df["responsetime"]))
    
#     filename = string(model_name,string(namb),"calls",string(ncalls),"batch",string(i),".csv")
#     open(filename, "w") do filename
#         writedlm(filename,  df["responsetime"])
#     end
#     print(filename)
# end

In [ ]:
"""all models simulation"""
# turnaround = Distributions.LogNormal(3.65, 0.3)
# lambda = 0
# local_path = ""
# model_name = "Stochastic"
# namb = 40
# ncalls = 1000
# #using AmbulanceDeployment

# pwd()

# cd(currentpath * "/test/")
# pwd()

# #using AmbulanceDeployment
# using Dates

# #include("..//src//dispatch/closestdispatch.jl")
# include("..//src//problem.jl")
# #include("..//src//evaluate.jl") #if you include, you need to include JuMP
# #include("..//src//simulate.jl")

# hospitals = CSV.File(string(local_path,"../test/austin-data/hospitals.csv")) |> DataFrame
# stations = CSV.File(string(local_path,"../test/austin-data/stations.csv")) |> DataFrame
# hourly_calls = CSV.File("../test/austin-data/Full_WeekdayCalls.csv") |> DataFrame
# adjacent_nbhd = CSV.File("../test/austin-data/adjacent_nbhd.csv") |> DataFrame
# coverage = CSV.read("../test/austin-data/coverage_real.csv", DataFrame, header=false)
# coverage = convert(Array{Bool, 2}, coverage[:, :])
# incidents = CSV.File("../test/austin-data/austin_incidents.csv") |> DataFrame
# solverstats = JLD.load("../src/outputs/austin_team_stats_1_03.jld")
# amb_deployment = solverstats["amb_deployment"]
# model_dict = Dict{String, Symbol}("Stochastic"=>:Stochastic, "Robust01"=>:Robust01, "Robust005"=>:Robust005, "Robust001"=>:Robust001, "Robust0001"=>:Robust0001,
#           "Robust00001"=>:Robust00001, "MEXCLP"=>:MEXCLP, "MALP"=>:MALP)


# p = DeploymentProblem(
#           hourly_calls,
#           adjacent_nbhd,
#           coverage,
#           namb = namb,
#           train_filter = (hourly_calls[!,:year] .== 2019) .* (hourly_calls[!,:month] .<= 3)
#           )
# test_calls = CSV.File("../test/austin-data/austin_test_calls_update1.csv")|> DataFrame 
# #update 1: remove negative indices
# test_calls = test_calls[1:ncalls,:]
# println("running $namb ambuances & $ncalls calls")

# # distribution of ambulances across all of the stations
# x = amb_deployment[model_dict[model_name]][namb]

# #establishes the dispatch problem
# problem = DispatchProblem(test_calls, hospitals, stations, p.coverage, x, turnaround=turnaround)

# #establishes the dispatch model
# dispatch = ClosestDispatch(p, problem)

# Random.seed!(1234)

In [65]:
println(mean(df["responsetime"]))
println(mean(df["scenetime"]))
println(mean(df["conveytime"]))
println(mean(df["returntime"]))

7.7477
16.250616666666666
24.4176
9.114916666666668


println(mean(df["responsetime"]))
println(mean(df["scenetime"]))
println(mean(df["conveytime"]))
println(mean(df["returntime"]))
#### lognromal 3.08
9.813166666666667 <br>
9.134116666666667 <br>
24.4176 <br>
8.908 <br>
#### lognromal 3.65
10.3443 <br>
16.000433333333337 <br>
24.4176 <br>
9.090300000000001 <br>
#### adjust travel time by time_bin weights
10.466350000000002 <br>
16.000433333333334 <br>
24.4176 <br>
9.073 <br>
#### global
7.7477 <br>
16.250616666666666 <br>
24.4176 <br>
9.114916666666668 <br>


In [ ]:
array = filter(x->typeof(x)==gui_event,guiArray)
responded_array = filter(x->x.event_type == "call responded", array)

In [ ]:
using DelimitedFiles
filename = string("visualization/time_adjust/",model_name,string(namb),"calls",string(ncalls),"stochastic40_timebins",".csv")
open(filename, "w") do filename
    writedlm(filename,  df["responsetime"])
end
print(filename)

In [ ]:
#using CSV
#filename = string(joshpath,"test/",model_name,string(namb),"calls",string(ncalls),".csv")
#CSV.write(filename, df, delim=",")